:pencil: philosophy documentation (#13)
FBerthelot/component-test-utils: website/docs/philosophy.md

title: Philosophy about component testing
---

## Philosophy of testing

Before introduce the way you can test your front-end application, we need to share the same definitions.

## Types of tests

> Component-test-util assume theses approximative definitions that fit well for component testing:

Reviewer: ```suggestion
Component-test-utils assumes these approximative definitions well-fitted for component testing:
```

> - _Unit Test_: Test that stay in memory. (No possible to access to HTTP, File System, ...)

Reviewer: ```suggestion
- _Unit Test_: Test that stay in memory. (No access to: HTTP, File System, ...)
```

> - _Integration test_: Test that can access to other things than memory (Network).

Reviewer: ```suggestion
- _Integration test_: Test that can access to other things than memory (Network, I/O).
```

- _End to End test_: It's basically an integration test with a special rule. No Mock or Stub allowed.

You can do both unit and integration test with component test-utils.

> The only test type you cannot do is end-to-end testing because component-test-util is basically a mock of your favorite framework.

Reviewer: ```suggestion
The only test type you cannot do is end-to-end testing because component-test-utils is basically a mock of your favorite framework.
```

### Component

No matter the JavaScript framework (Vanilla, Vue.js, Angular, React, ...) you use, they share some common vision of what a component should be.

> A component is a Custom HTML element that produce a view (aka. a piece of others Component and HTML elements).

Reviewer: ```suggestion
A component is a Custom HTML element that produces a view (aka. a piece of others Component and HTML elements).
```

![Component that generate view](/img/component_view.gif)

> This component can received parameter just like a function. Usually those parameters are called properties or props.

Reviewer: ```suggestion
This component can receive parameters just like a function. Usually those parameters are called properties or props.
```

![Component that generate view according props](/img/component_props.gif)

Those props can produce a different view according to the props value.

> So no matter the framework you use, the easiest to write is to test how the view is rendered according to the props you give.

Reviewer: ```suggestion
As the render only depends on the props, no matter the framework you use, the easiest is to write tests to assert the view is rendered according to the props you give.
```

> This can easily achieved with component-test-utils. For example, if you consider testing a like button:

Reviewer: ```suggestion
This can easily be achieved with component-test-utils. For example, if you consider testing a like button:
```

```js
describe('like button component', () => {
  it('should render a button with "1 like" when props give 1 like', () => {
    const component = shallow(Component, {props: {nbLikes: 1}});

    expect(component.html()).toContain('1 like');
  });

  it('should render a button with "2 likes" when props give 2 likes', () => {
    const component = shallow(Component, {props: {nbLikes: 2}});

    expect(component.html()).toContain('2 likes');
  });
});
```

### Statefull component

> Unfortunately, component is a bit more complex sometimes. Sometimes component have is own state.

Reviewer: ```suggestion
Unfortunately, components can sometimes be a bit more complex. For example when they have their own state.
```

> Let's taking the previous example, imagine the like button as different style if user have liked and if it's not.

Reviewer: ```suggestion
Considering the previous example, imagine the like button has different styles depending on whether or not the user have liked the content related.
```

> As component-test-utils consider component as a black box, you cannot modify or access internal state of your component.

Reviewer: ```suggestion
As component-test-utils consider component as a black box, you cannot modify or access its internal state.
```

> Instead, you have to trigger event from the view itself !

Reviewer: ```suggestion
Instead, you have to trigger an event from the view itself !
```

![Statefull component](/img/component_state.gif)

> For example, if you want testing the different styling of your like button :

Reviewer: ```suggestion
For example, if you want to test the different styles of your like button :
```

```javascript
describe('like button style', () => {
  it('should set button to notLiked by default', () => {
    const component = shallow(Component, {props: {nbLikes: 1}});

    component.querySelector('button').dispatchEvent('click');

    expect(component.html().querySelector('button').props.class).not.toContain(
      'liked'
    );
  });

  it('should set button to liked when clicking on the button', () => {
    const component = shallow(Component, {props: {nbLikes: 1}});

    component.querySelector('button').dispatchEvent('click');

>     expect(component.html().querySelector('button').props.class).not.toContain(

Reviewer: ```suggestion
    expect(component.html().querySelector('button').props.class).toContain(
```

      'liked'
    );
  });
});
```

### Component can emit event

> **Important: As only react is supported right now, theses feature are not yet implemented because react don't need it**

Reviewer: ```suggestion
**Important: React being the only rendering library supported for now, these features are not implemented yet as they're not needed in react**
```

> As component

Reviewer: Given the component below ?

![Component that emit event](/img/component_event.png)

```javascript
describe('like button - onLike event', () => {
  it('should emit onClick Event', () => {
    const spy = createSpy();

>     // const spy = jest.fn(); on jest

Reviewer: ```suggestion
    // const spy = jest.fn(); using jest
```

    const component = shallow(Component, {events: {onLike}});

    component.querySelector('button').dispatchEvent('click');

    expect(spy).toHaveBeenCalled();
  });
});
```

> ### Component sometimes have more than props

Reviewer: ```suggestion
### Component with multiple props
```

> **Important: As only react is supported right now, theses feature are not yet implemented because react don't need it**

Reviewer: same suggestion as above :bulb:

For example, Angular component can inject services in the constructor of the components.

> In order to provide a standardized way external object that can affect the rendered view other than props, a external option is available.

Reviewer: ```suggestion
In order to provide a standardized way an external object can affect the rendered view while not being considered a property, a external option is available.
```
Unsure what you mean here 🤔 :

![Component with externals](/img/component_externals.png)

An example :

```javascript
describe('user component', () => {
  it('should display the user got from the service', () => {
    const component = shallow(Component, {
      externals: {
        userService: {
          getUser: () => Promise.resolve({name: 'component-test-utils'})
        }
      }
    });

    expect(component).toContain('component-test-utils');
  });
});
```

### Mocking strategies

> Component doesn't only generate HTML element, they can have sub-components in their view. For this kind of "parent" component, the main question of the developer should be: "should I mock this child component ?".

Reviewer: ```suggestion
Component doesn't only generate HTML element, they can have sub-components in their view. For this kind of "parent" component, this is the main question the developer should ask: "should I mock this child component ?".
```

> For this use case, Component-test-utils have two way to create a component in an test environment:

Reviewer: ```suggestion
For this use case, Component-test-utils has two ways to create a component in a test environment:
```

> - _White list_: Every sub-component is mocked, you can give a list a component that won't be mocked and give their mock.

Reviewer: ```suggestion
- _White list_: Every sub-component is mocked automatically and you can specify the mock for each component by providing a map.
```

Reviewer: I hope I understood correctly what you meant.
Maybe an example could be useful here, at least a philosophical one 💡

    expect(component.html().querySelector('button').props.class).not.toContain(
      'liked'
    );
  });

  it('should set button to liked when clicking on the button', () => {
    const component = shallow(Component, {props: {nbLikes: 1}});

    component.querySelector('button').dispatchEvent('click');

    expect(component.html().querySelector('button').props.class).toContain(
      'liked'
    );
  });
});
```

### Component, event output

**Important: React being the only rendering library supported for now, these features are not implemented yet as they're not needed in react**

Given the component sometimes need to talk with parent component, frameworks use event to achieved that.

![Component that emit event](/img/component_event.png)

To ensure your component emit the good event, you can give spies to components and ensure they are called.

```javascript
describe('like button - onLike event', () => {
  it('should emit onClick Event', () => {
    const spy = createSpy();
    // const spy = jest.fn(); using jest
    const component = shallow(Component, {events: {onLike}});

    component.querySelector('button').dispatchEvent('click');

>     expect(spy).toHaveBeenCalled();

Reviewer 2: Alors ok, je sais que c'est du pseudo code, mais tester du code asynchrone avec une syntaxe synchrone, ça pique un peux les yeux.

Author: #20

  });
});
```

### Component with multiple props

> **Important: React being the only rendering library supported for now, these features are not implemented yet as they're not needed in react**

Reviewer 2: ```suggestion
**Important: React being the only rendering library currently supported, these features are not implemented yet as they're not needed in react**
```
